File: 최한준/2주차[백트래킹]/14888.py
import sys
import math
get_line: iter = lambda: map(int, sys.stdin.readline().rstrip().split())
get_input: int = lambda: int(sys.stdin.readline().strip())

min_value = None
max_value = None
N = None
arr = None
op_cnt = None
op_arr = None
cnt = 0

def set_variable():
    global min_value, max_value, N, arr, op_cnt, op_arr
    N = get_input()
    arr = list(get_line())
    op_arr = []
    op_cnt = list(get_line())
    min_value = math.inf
    max_value = -math.inf # 알게된 것.

def solution():
    global min_value, max_value, N, arr, op_cnt, op_arr,cnt
    def cal():
        ret = arr[0]
        for num, op in zip(arr[1:], op_arr):
            if op   == 0 : ret += num
            elif op == 1 : ret -= num
            elif op == 2 : ret *= num
            elif op == 3 : ret = int(ret / num)
        return ret
    
    def rec(depth):
        if depth == len(arr) - 1:
            max_value = max(max_value, cal())
            min_value = min(min_value, cal())
            return
        else:
            for idx in range(4):
                if op_cnt[idx] > 0:
                    op_cnt[idx] -= 1
                    op_arr.append(idx)
                    rec(depth + 1)
                    op_arr.pop()
                    op_cnt[idx] += 1
                    
    
    rec(0)
    print(max_value)
    print(min_value)


if __name__ == "__main__":
    set_variable()
    solution()

File: 양진기/week2[백트래킹]/14888.java
import java.util.*;
import java.io.*;

class Main{

    static int N;
    static int[] arr;
    static int[] operator = new int[4];
    static int MAX = Integer.MIN_VALUE;
    static int MIN = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        arr = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for(int i=0; i<N; i++){

            arr[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for(int i=0; i<4; i++){
            operator[i] = Integer.parseInt(st.nextToken());
        }

        cal(arr[0], 1);

        System.out.println(MAX);
        System.out.println(MIN);
    }

    static void cal(int num, int idx){

        if(idx == N){
            MAX = Math.max(MAX, num);
            MIN = Math.min(MIN, num);
            return;
        }

        for(int i=0; i<4; i++){
            // 연산자 개수가 1개 이상인 경우
            if(operator[i] > 0){

                // 해당 연산자를 1 감소시킨다.
                operator[i]--;

                switch(i){
                    case 0: cal(num + arr[idx], idx + 1);    break;
                    case 1: cal(num - arr[idx], idx + 1);    break;
                    case 2: cal(num * arr[idx], idx + 1);    break;
                    case 3: cal(num / arr[idx], idx + 1);    break;
                }

                // 재귀호출이 종료되면 다시 해당 연산자 개수를 복구한다.
                operator[i]++;
            }
        }
    }
}